### DISCUSSION TITLE
have you seen ProjecturEd?

### DISCUSSION THREAD

**Discussion by attila-lendvai** [neomacs-project/neomacs#56] [General]
just making sure you're aware of this project: https://github.com/projectured/projectured/

**Comment by kchanqvq** [comment]
Thanks for bringing this up! I came across this but haven't try it myself. I'd love to see if there're things I can learn from!

Some problems that I'm currently trying to figure out:
- How to copy and paste between different modes? They might require different structure
- How to keep a valid structure in any mode with a general editing facility? Modes have different requirement on the data it contains (expecting different node types, parent-child relations, etc). Past HTML/XML editor seems to suggest some kind of schema system.

Do you have any idea if ProjecturEd has tackled these problems?

**Comment by attila-lendvai** [comment]
re validity: whether a strucutre is valid or not depends on who is looking at it and with what expectations.

e.g. a Java function call with a png image as first argument may be invalid for the java compiler (as-is), but valid for the editor that can happily display and modify that structure with the image displayed inline.

i believe this kinda answers the copy paste question also: you can edit the structure as you see fit. of course metadata can guide the editing, and an interactive ping-pong between the domain tools (e.g. a java compiler) and the editing framework and its display can be a useful help in pointing out how the structure is not appropriate for the compiler, or whatever pipeline is consuming it.

**Reply by kchanqvq** [reply]
I think there's more to it. The above solves the problem that different tools (outside the editor) might expect different formats (which Neomacs also supports via `write-dom-aux` and `read-dom-aux` generic functions). But there's still problem that different **editors** expect different kind of structure. For example, the Java editor (in Neomacs's case, `lisp-mode`), might not be able to handle image at all, or at least does a bad job at it. I think this is inevitable because I want `lisp-mode` to be really good at editing Lisp. A single unified editor would solve this, but I don't think it's possible to make one that's good at everything. 

I created an issue at #84. Love to hear more from a different perspective if you have more ideas!

**Reply by attila-lendvai** [reply]
what if the dispatching on type/context/language happens on every node?

e.g. if a node of type `lisp-ast` is found, then `lisp-mode` is called upon handle the editing and display. but if at argument 3 of a function call there's a PNG image, then the image handler is called upon that when rendering to screen renders a thumbnail, and when preparing-to-compile then saves the image as a resource and inserts the correct call to look up that resource (which may depend on the target of the compilation).

**Reply by kchanqvq** [reply]
That IS a really good idea! I thought about this before as well, "node-local" modes. However I couldn't figure out all the details, e.g.: how to handle cursor leaving and entering nodes with different modes -- who is going to decide where the cursor would land, say for a `M-f` or `C-M-f`? There are many more operations which seem to have strong enough coupling between any nodes inside the same buffer as well. Therefore I started writing Neomacs with only buffer level modes. Although now I have gained more experience with a working structural editor I might be able to rethink this and move towards this direction, e.g.: the cursor moving example can probably be solved with the `selectable-p-aux` generic function which let each node compute selectable positions inside it.

**Comment by attila-lendvai** [comment]
there was an interview recorded with ProjecturEd's author. it's in the post-processing phase, and hopefully published soon.

i think it will present a bird's eye view of the internal model of PrEd.

**Reply by kchanqvq** [reply]
Looking forward to it! Will there be a transcript or slide? I process text better than videos...
